demos/sample_azure_iot_pnp/sample_azure_iot_pnp.c

```diff
@@ -0,0 +1,1062 @@
+                                              uint32_t * ulOutVersion )
+{
+    AzureIoTResult_t xResult;
+    AzureIoTJSONReader_t xReader;
+    AzureIoTJSONTokenType_t xTokenType;
+    const uint8_t * pucComponentName = NULL;
+    uint32_t ulComponentNameLength = 0;
+
+    *xOutTemperature = 0.0;
+
+    xResult = AzureIoTJSONReader_Init( &xReader, pxMessage->pvMessagePayload, pxMessage->ulPayloadLength );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTHubClientProperties_GetPropertiesVersion( &xAzureIoTHubClient, &xReader, pxMessage->xMessageType, ulOutVersion );
+
+    if( xResult != eAzureIoTSuccess )
+    {
+        LogError( ( "Error getting the property version" ) );
+    }
+    else
+    {
+        /* Reset JSON reader to the beginning */
+        xResult = AzureIoTJSONReader_Init( &xReader, pxMessage->pvMessagePayload, pxMessage->ulPayloadLength );
+        configASSERT( xResult == eAzureIoTSuccess );
+
+        while( ( xResult = AzureIoTHubClientProperties_GetNextComponentProperty( &xAzureIoTHubClient, &xReader,
+                                                                                 pxMessage->xMessageType, xPropertyType,
+                                                                                 &pucComponentName, &ulComponentNameLength ) ) == eAzureIoTSuccess )
+        {
+            xResult = AzureIoTJSONReader_TokenType( &xReader, &xTokenType );
+            configASSERT( xResult == eAzureIoTSuccess );
+
+            if( AzureIoTJSONReader_TokenIsTextEqual( &xReader,
+                                                     sampleazureiotPROPERTY_TARGET_TEMPERATURE_TEXT,
+                                                     strlen( sampleazureiotPROPERTY_TARGET_TEMPERATURE_TEXT ) ) )
+            {
+                xResult = AzureIoTJSONReader_NextToken( &xReader );
+                configASSERT( xResult == eAzureIoTSuccess );
+
+                xResult = AzureIoTJSONReader_TokenType( &xReader, &xTokenType );
+                configASSERT( xResult == eAzureIoTSuccess );
+
+                if( xTokenType != eAzureIoTJSONTokenNUMBER )
+                {
+                    LogError( ( "Invalid value for target temperature reached" ) );
+                    xResult = eAzureIoTErrorFailed;
+                    break;
+                }
+
+                /* Get desired temperature */
+                xResult = AzureIoTJSONReader_GetTokenDouble( &xReader, xOutTemperature );
+                configASSERT( xResult == eAzureIoTSuccess );
+
+                xResult = AzureIoTJSONReader_NextToken( &xReader );
+                configASSERT( xResult == eAzureIoTSuccess );
+            }
+            else
+            {
+                xResult = AzureIoTJSONReader_NextToken( &xReader );
+                configASSERT( xResult == eAzureIoTSuccess );
+
+                xResult = AzureIoTJSONReader_SkipChildren( &xReader );
+                configASSERT( xResult == eAzureIoTSuccess );
```

Comment:
we are assuming that next object is complex object.

how about using this impl:

[https://github.com/azure-rtos/netxduo/blob/ab1b3706e31f4ca9701b6c215a143386b1b655f9/addons/azure_iot/samples/sample_azure_iot_embedded_sdk_pnp.c#L328](https://github.com/azure-rtos/netxduo/blob/ab1b3706e31f4ca9701b6c215a143386b1b655f9/addons/azure_iot/samples/sample_azure_iot_embedded_sdk_pnp.c#L328)

Reply:
It's not so much an assumption as much as it is covering the case when it is a complex object. It's a no-op if it isn't a complex object. Same thing is done here:
https://github.com/Azure/azure-sdk-for-c/blob/75afd98a1efd41d0c1f4475f692da88547a61582/sdk/samples/iot/paho_iot_pnp_component_sample.c#L507-L509